FirebaseSegmentation/Sources/SEGDatabaseManager.m

```diff
@@ -0,0 +1,419 @@
+      kSEGAssociationStatusKey : associationStatus
+    };
+    [associations setObject:associationData forKey:firebaseApplicationName];
+  }
+  sqlite3_finalize(statement);
+  return associations;
+}
+
+/// Returns the current version of the Remote Config database.
++ (NSString *)pathForSegmentationDatabase {
+  NSArray<NSString *> *dirPaths =
+      NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
+  NSString *appSupportPath = dirPaths.firstObject;
+  NSArray<NSString *> *components =
+      @[ appSupportPath, kApplicationSupportSubDirectory, kDatabaseName ];
+  return [NSString pathWithComponents:components];
+}
+
+- (BOOL)createTableSchema {
+  SEG_MUST_NOT_BE_MAIN_THREAD();
+  NSString *mainTableSchema =
+      [NSString stringWithFormat:@"create TABLE IF NOT EXISTS %@ (_id INTEGER PRIMARY KEY, %@ "
+                                 @"TEXT, %@ TEXT, %@ TEXT, %@ TEXT)",
+                                 kMainTableName, kMainTableColumnApplicationIdentifier,
+                                 kMainTableColumnCustomInstallationIdentifier,
+                                 kMainTableColumnFirebaseInstallationIdentifier,
+                                 kMainTableColumnAssociationStatus];
+
+  return [self executeQuery:[mainTableSchema cStringUsingEncoding:NSUTF8StringEncoding]];
+}
+
+- (void)removeDatabase:(NSString *)path {
+  SEG_MUST_NOT_BE_MAIN_THREAD();
+  if (sqlite3_close(self->_database) != SQLITE_OK) {
+    [self logDatabaseError];
+  }
+  self->_database = nil;
+
+  NSFileManager *fileManager = [NSFileManager defaultManager];
+  NSError *error;
+  if (![fileManager removeItemAtPath:path error:&error]) {
+    FIRLogError(kFIRLoggerSegmentation, @"I-SEG000011",
+                @"Failed to remove database at path %@ for error %@.", path, error);
+  }
+}
+
+#pragma mark - execute
+- (BOOL)executeQuery:(const char *)SQL {
+  SEG_MUST_NOT_BE_MAIN_THREAD();
+  char *error;
+  if (sqlite3_exec(_database, SQL, nil, nil, &error) != SQLITE_OK) {
+    FIRLogError(kFIRLoggerSegmentation, @"I-SEG000012", @"Failed to execute query with error %s.",
+                error);
+    return NO;
+  }
+  return YES;
+}
+
+#pragma mark - insert
+- (void)insertMainTableApplicationNamed:(NSString *)firebaseApplication
+               customInstanceIdentifier:(NSString *)customInstanceIdentifier
+             firebaseInstanceIdentifier:(NSString *)firebaseInstanceIdentifier
+                      associationStatus:(NSString *)associationStatus
+                      completionHandler:(SEGRequestCompletion)handler {
+  // TODO: delete the row first.
+  __weak SEGDatabaseManager *weakSelf = self;
+  dispatch_async(_databaseOperationQueue, ^{
+    NSArray<NSString *> *values =
+        [[NSArray alloc] initWithObjects:firebaseApplication, customInstanceIdentifier,
+                                         firebaseInstanceIdentifier, associationStatus, nil];
```

Comment:
Use an array literal instead of `initWithObjects:`.

```objc
NSArray<NSString *> *values = @[
  firebaseApplication, customInstanceIdentifier, firebaseInstanceIdentifier, associationStatus
];
```